### PR TITLE
Bugfix: button with text and icon is recognized as icon only

### LINKED ISSUE #928: [BUG] Button with text and icon is recognized as icon only
**Describe the bug**
When creating a kirby button with text first and a kirby icon second and then adds a ngIf statement to the icon, so it only should show when a condition is met, the button is recognized as an icon only button. 

**To Reproduce**

1. Create a kirby button
2. Add button text inside
3. Add a kirby icon inside
4. Apply a ngIf statement with a condition
5. See how the kirby button is rendered in a wrong way, the frame is too small as it is meant  for a button with only an icon and the button text is not fully visible.

**Kirby version**

0.8.101

**Expected behavior**
A kirby button with text to the left and a kirby icon to the right, which is only shown if the specific condition is met.

**Screenshots**
Skærmbillede 2020-02-24 kl  10 52 09

**Please complete the following information:**

- OS: MacOS
- Browser: Chrome
- Version 80.0.3987.116 

**Additional context**

*Solve*
- [ ] Move issue from "Ready to do" -> "in progress" on our [project board](https://github.com/kirbydesign/designsystem/projects/1)
- [ ] Create Fix Branch
- [ ] Create test reproducing the error
- [ ] Add any required mock and fakes to the `KirbyTestingModule`
- [ ] Fix
- [ ] If relevant [Update Examples and Showcase](https://cookbook.kirby.design/home/showcase/button)
- [ ] Push commits to GitHub
- [ ] Open a pull request (remember to add `closes #issueno` to the description) and ask @kirbydesign/kirby-guild for review
- Merge to master and delete branch

### PR DESCRIPTION
**Pull request by MadsBuchmann**
This PR closes #928.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #928

## What is the new behavior?
The following case where a kirby-icon with a ngIf directive is inserted before the button text: 
billede

Will no longer result in the button being rendered as a button with icon only: 
billede

It will instead be rendered as a button with icon and text. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by MadsBuchmann** [comment]
The `nativeElement` local variable in `findNonCommentSibling` is typed as any, due to the line ` const iconNativeElement = this.iconDomNode.nativeElement;` (button.component.ts - line 73) resulting in `iconNativeElement` having the type `any`.

File: libs/designsystem/src/lib/components/button/button.component.spec.ts [hunk 1]
```diff
@@ -313,3 +314,29 @@ describe('ButtonComponent', () => {
     });
   });
 });
+
+describe('ButtonComponent using an ngIf directive', () => {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/button/button.component.spec.ts:318]: Consider moving to [existing button integration test](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/button/button.component.integration.spec.ts#L466)

File: libs/designsystem/src/lib/components/button/button.component.ts [hunk 2]
```diff
@@ -71,18 +70,27 @@ export class ButtonComponent implements AfterContentInit {
...
-      if (next && next.nodeName !== '#comment') {
+      if (next) {
         this._hasSlottedContent = true;
         this._isIconLeft = true;
         this._isIconRight = false;
       }
     }
   }
+
+  private findNonCommentSibling(nativeElement: any, direction: 'nextSibling' | 'previousSibling') {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/button/button.component.ts:89]: You can strong type the `nativeElement` parameter by typing the `iconDomNode` member var as:
 `iconDomNode: ElementRef<HTMLElement>;`

File: libs/designsystem/src/lib/components/button/button.component.ts [hunk 3]
```diff
@@ -71,18 +70,27 @@ export class ButtonComponent implements AfterContentInit {
 
   ngAfterContentInit(): void {
     if (this.iconDomNode && this.iconDomNode.nativeElement) {
-      const prev = this.iconDomNode.nativeElement.previousSibling;
-      const next = this.iconDomNode.nativeElement.nextSibling;
-      if (prev && prev.nodeName !== '#comment') {
+      const iconNativeElement = this.iconDomNode.nativeElement;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/button/button.component.ts:73]: Consider aligning naming with similar code in repo:
```suggestion
      const iconElement = this.iconElementRef.nativeElement;
```

**Review by jakobe — CHANGES_REQUESTED** [review]
Nice work! 👍
Only a few suggestions 😄

File: libs/designsystem/src/lib/components/button/button.component.spec.ts (hunk 1 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.spec.ts]: Should be resolved by new commits.

File: libs/designsystem/src/lib/components/button/button.component.ts (hunk 2 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.ts]: Done! - however had to cast the `sibling` variable in `findNonCommentSibling` to a `HTMLElement` for it to not complain. I personally don't think this is an issue in this context, as it is happening inside the function, and is "hidden" from the consumer. 

But wouldn't mind your toughts on it 😄

File: libs/designsystem/src/lib/components/button/button.component.ts (hunk 3 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.ts]: Should be resolved by new commits.

File: libs/designsystem/src/lib/components/button/button.component.integration.spec.ts [hunk 4]
```diff
@@ -589,3 +590,29 @@ describe('ButtonComponent configured with text and icon', () => {
     });
   });
 });
+
+describe('ButtonComponent using an ngIf directive', () => {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/button/button.component.integration.spec.ts:594]: Please align naming:
```suggestion
describe('ButtonComponent configured with text and icon using an ngIf directive', () => {
```

File: libs/designsystem/src/lib/components/button/button.component.ts [hunk 5]
```diff
@@ -86,8 +86,11 @@ export class ButtonComponent implements AfterContentInit {
     }
   }
 
-  private findNonCommentSibling(nativeElement: any, direction: 'nextSibling' | 'previousSibling') {
-    const sibling = nativeElement[direction];
+  private findNonCommentSibling(
+    nativeElement: HTMLElement,
+    direction: 'nextSibling' | 'previousSibling'
+  ) {
+    const sibling = nativeElement[direction] as HTMLElement;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/button/button.component.ts:93]: Consider using direction to conditionally get sibling instead:
```suggestion
    direction: 'previous' | 'next'
  ) {
    const sibling = direction === 'previous' ? nativeElement.previousSibling : nativeElement.nextSibling;
```

File: libs/designsystem/src/lib/components/button/button.component.ts [hunk 6]
```diff
@@ -86,8 +86,11 @@ export class ButtonComponent implements AfterContentInit {
     }
   }
 
-  private findNonCommentSibling(nativeElement: any, direction: 'nextSibling' | 'previousSibling') {
-    const sibling = nativeElement[direction];
+  private findNonCommentSibling(
+    nativeElement: HTMLElement,
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/button/button.component.ts:90]: Consider naming
```suggestion
    element: HTMLElement,
```

**Review by jakobe — CHANGES_REQUESTED** [review]

File: libs/designsystem/src/lib/components/button/button.component.ts (hunk 6 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.ts]: Should be resolved by new commits 😊

File: libs/designsystem/src/lib/components/button/button.component.ts (hunk 5 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.ts]: Should be resolved by new commits 😊

File: libs/designsystem/src/lib/components/button/button.component.integration.spec.ts (hunk 4 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.integration.spec.ts]: Should be resolved by new commits 😊

File: libs/designsystem/src/lib/components/button/button.component.ts (hunk 2 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.ts]: note: typecast has been moved as part of the argument when doing the recursive call to `findNonCommentSibling`.

File: libs/designsystem/src/lib/components/button/button.component.ts [hunk 7]
```diff
@@ -66,23 +65,32 @@ export class ButtonComponent implements AfterContentInit {
...
         this._isIconLeft = true;
         this._isIconRight = false;
       }
     }
   }
+
+  private findNonCommentSibling(element: HTMLElement, direction: 'next' | 'previous') {
+    const sibling = direction === 'previous' ? element.previousSibling : element.nextSibling;
+    if (sibling != null && sibling.nodeName === '#comment') {
+      return this.findNonCommentSibling(sibling as HTMLElement, direction);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/button/button.component.ts:92]: The cast can be removed by generalizing the paramater type: 
```suggestion
  private findNonCommentSibling(element: Element, direction: 'next' | 'previous') {
    const sibling = direction === 'previous' ? element.previousSibling : element.nextSibling;
    if (sibling != null && sibling.nodeName === '#comment') {
      return this.findNonCommentSibling(sibling, direction);
```

**Review by jakobe — CHANGES_REQUESTED** [review]

**Review by jakobe — APPROVED** [review]

File: libs/designsystem/src/lib/components/button/button.component.ts [hunk 8]
```diff
@@ -66,23 +65,32 @@ export class ButtonComponent implements AfterContentInit {
 
   @ContentChild(IconComponent, { static: false }) icon: IconComponent;
   @ContentChild(IconComponent, { static: false, read: ElementRef })
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/button/button.component.ts]: Should be resolved by ac49b07

**Review by jakobe — APPROVED** [review]
giphub

**Pull request merged**